src/minitrace_grpc.rs

```diff
@@ -0,0 +1,106 @@
+    S::Future: Send + 'static,
+{
+    type Response = S::Response;
+    type Error = S::Error;
+    type Future = futures::future::BoxFuture<'static, Result<Self::Response, Self::Error>>;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx)
+    }
+
+    fn call(&mut self, req: tonic::codegen::http::Request<Body>) -> Self::Future {
+        // This is necessary because tonic internally uses `tower::buffer::Buffer`.
+        // See https://github.com/tower-rs/tower/issues/547#issuecomment-767629149
+        // for details on why this is necessary
+        let clone = self.inner.clone();
+        let mut inner = std::mem::replace(&mut self.inner, clone);
+
+        let sender = self.sender.clone();
+
+        let headers = req.headers().clone();
+
+        Box::pin(async move {
+            let (root_span, collector) = Span::root("service_root");
+
+            let response = inner.call(req).in_span(root_span).await?;
+
+            let spans = collector.collect();
+
+            let rand: u32 = rand::random(); // todo: what is an appropriate default span id if not present in headers, uuid? something other than random number?
+            let default_trace_id_header_value = HeaderValue::from_str(&rand.to_string()).unwrap();
+            let trace_id_header = headers.get("x-datadog-trace-id")
+                .unwrap_or(&default_trace_id_header_value).to_str();
+            let trace_id: u64 = trace_id_header.unwrap().parse::<u64>().unwrap();
+            let span_id_prefix: u32 = 0;
+            let default_parent_span_id_header_value = HeaderValue::from_str(&rand.to_string()).unwrap();
+            let parent_span_id_header = headers.get("x-datadog-parent-id")
+                .unwrap_or(&default_parent_span_id_header_value).to_str();
+            let parent_span_id: u64 = parent_span_id_header.unwrap().parse::<u64>().unwrap();
+
+            sender.send(MinitraceSpans {
+                trace_id,
+                parent_span_id,
+                span_id_prefix,
+                spans: Box::new(spans)
+            }).expect("Failed to send spans to channel");
+
+            Ok(response)
+        })
+    }
+}
+
+pub struct MinitraceSpans {
+    trace_id: u64,
+    parent_span_id: u64,
+    span_id_prefix: u32,
+    spans: Box<Vec<minitrace::span::Span>>
+}
+
+pub async fn report_datadog_traces(receiver: Receiver<MinitraceSpans>, host: IpAddr, port: u16, service_name: String) {
+    let socket = SocketAddr::new(host, port);
+    loop {
+        let spans = receiver.recv().unwrap();
+
+        let bytes = Reporter::encode(
+            &service_name,
+            spans.trace_id,
+            spans.parent_span_id,
+            spans.span_id_prefix,
+            &*spans.spans,
+        )
+        .expect("encode error");
+        Reporter::report_blocking(socket, bytes).expect("report error");
```

Comment:
We will want to make this nonblocking or write our own if they don't provide it.

Reply:
Looks like there's a non-blocking version here: https://github.com/tikv/minitrace-rust/blob/master/crates/minitrace-datadog/src/lib.rs#L67

Reply:
If the task spawns on a different thread it might not be an issue, but I think the default runtime determines which tasks are assigned to which threads arbitrarily. ` #[tokio::main]` builds a multi-threaded runtime by default.

Reply:
Yeah, I'll look at the non-blocking version. I tried that at first and I think I either ran into some problem or was just confused about something around it, and figured it didn't matter since I was thinking about this as if it was its own thread for sure

Reply:
So, when I try the non-blocking version I end up getting this error ```thread 'tokio-runtime-worker' panicked at 'there is no reactor running, must be called from the context of a Tokio 0.2.x runtime'```...

I see in this library's Cargo.toml file that there is a dev dependency on Tokio 0.2, but I would think that dev dependencies shouldn't affect me, is that right? Do I need to build with some prod flag or something?

Reply:
Just saw this I'm not really sure what would cause that. It wouldn't be too bad to just implement that function locally. There's not much going on in there.

Reply:
I was experimenting yesterday with just sending these via a UdpSocket, and I could definitely get stuff sent to my local DatadogAgent, but I'm having trouble finding a reference for what the format of spans over that udp interface are supposed to be, everything I can find in the datadog docs around message format seems to be specific to other metric collection, which is more of a simple format like ```<METRIC_NAME>:<VALUE>|<TYPE>|@<SAMPLE_RATE>|#<TAG_KEY_1>:<TAG_VALUE_1>,<TAG_2>```

Reply:
I just switched to using reqwest directly, and after much digging, I think the udp socket doesn't support the tracing functionality, only the simpler metrics... at least none of my experiments worked and that is all that is referenced in the datadog docs, so we might be stuck with tcp 🤷